packages/core/contracts/zodiac/OptimisticGovernor.sol

```diff
@@ -0,0 +1,321 @@
+
+        // Add transactions to proposal in memory.
+        for (uint256 i = 0; i < _transactions.length; i++) {
+            require(_transactions[i].to != address(0), "The `to` address cannot be 0x0");
+            // If the transaction has any data with it the recipient must be a contract, not an EOA.
+            if (_transactions[i].data.length > 0) {
+                require(_isContract(_transactions[i].to), "EOA can't accept tx with data");
+            }
+        }
+        proposal.transactions = _transactions;
+
+        // proposalHashes[id] = keccak256(abi.encodePacked(proposalData));
+        proposalHashes[id] = keccak256(abi.encode(_transactions));
+
+        // Get the bond from the proposer and approve the bond and final fee to be used by the oracle.
+        uint256 totalBond = finalFee + bond;
+        collateral.safeTransferFrom(msg.sender, address(this), totalBond);
+        collateral.safeIncreaseAllowance(address(skinnyOptimisticOracle), totalBond);
+
+        // Propose a set of transactions to the OO. If not disputed, they can be executed with executeProposal().
+        // docs: https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/oracle/interfaces/SkinnyOptimisticOracleInterface.sol
+        skinnyOptimisticOracle.requestAndProposePriceFor(
+            identifier,
+            uint32(time),
+            ancillaryData,
+            collateral,
+            0,
+            bond,
+            uint256(liveness),
+            address(this),
+            // Canonical value representing "True"; i.e. the transactions are valid.
+            int256(1e18)
+        );
+
+        emit TransactionsProposed(id, proposer, time, proposal);
+    }
+
+    function executeProposal(
+        uint256 _proposalId,
+        Transaction[] memory _transactions,
+        bytes memory _explanation,
+        uint32 _originalTime,
+        SkinnyOptimisticOracleInterface.Request memory _request
+    ) public payable nonReentrant {
+        // Recreate the proposal hash from the inputs and check that it matches the stored proposal hash.
+        uint256 id = _proposalId;
+
+        // Construct the ancillary data.
+        bytes memory ancillaryData = _explanation;
+        AncillaryData.appendKeyValueUint(ancillaryData, "id", id);
+
+        // This will reject the transaction if the proposal hash generated from the inputs does not match the stored proposal hash.
+        // require(proposalHashes[id] == keccak256(abi.encodePacked(proposalData)), "proposal hash does not match");
+        require(proposalHashes[id] == keccak256(abi.encode(_transactions)), "proposal hash does not match");
+
+        // Remove proposal hash so transactions can not be executed again.
+        proposalHashes[id] = 0;
+
+        // This will revert if the price has not settled.
+        (, int256 price) =
+            skinnyOptimisticOracle.settle(address(this), identifier, _originalTime, ancillaryData, _request);
```

Comment:
This is where the skinny OO gets a little tricky.

If this is already settled, this will revert, meaning a 3rd party could block executions by pre-settling.

@nicholaspai could probably advise better than I can, but my suggestion would be:
1. Check _request.settled.
2. If true, call require(hasPrice()).
3. If false, call settle().

Same goes for the other callsite.

As a side note (that can be left for a follow-up): why did we decide to use the skinny OO here? It has considerable UX tradeoffs when asking 3rd parties to propose. Do we have any concerns around that?